### DISCUSSION TITLE
What are pnpm’s preferences for Node.js distribution?

### DISCUSSION THREAD

**Discussion by GeoffreyBooth** [pnpm/pnpm#7736] [General]
Forgive me if this was asked elsewhere; I’ve been looking and I can’t seem to find it. There’s an [open PR](https://github.com/nodejs/node/pull/51886) for Node to include a `pnpm` binary in its distribution which would be a runtime to Corepack, to tell Corepack to download and manage pnpm. Is this something that the pnpm project wants?

As I see it, there are three broad options:

1. Node.js doesn’t distribute pnpm at all: neither the pnpm tool itself nor a way to download it on demand. Users would need to download and install pnpm directly via one of the methods that pnpm recommends.
2. When users install Node, along with creating `node` and `npm` and `npx` binaries/symlinks, the installation creates a `pnpm` symlink to Corepack. When users run `pnpm` for the first time after Node installation, Corepack downloads and version-manages pnpm.
3. When users install Node, it creates a `pnpm` binary or script that does something else, like `npm install --global pnpm` or whatever installation method the pnpm team prefers.

Which of the above is the preference of the pnpm project? cc @zkochan

**Comment by trivikr** [comment]
> Is this something that the pnpm project wants?

pnpm recommends [using corepack](https://pnpm.io/installation#using-corepack) in their Getting Started page, as the first option after standalone scripts, and before npm.

**Reply by GeoffreyBooth** [reply]
So are you saying that you prefer option 2 or option 3?

For option 3 I meant `npm install` as an example. It could be any command, like it could be whatever standalone scripts command the pnpm team suggests.

**Reply by trivikr** [reply]
I primarily use `yarn` with corepack, and landed on this discussion from https://github.com/nodejs/node/pull/51886#issuecomment-1977584750

I quickly looked at pnpm Getting Started wiki, and highlighted that they also recommend corepack.

**Reply by GeoffreyBooth** [reply]
Well it’s one of 14 or so options, depending how you count, and there’s no text for any of them saying that any in particular are recommended or disfavored.

Regardless of how pnpm recommends their software be installed or _if_ they have a recommendation, the question is what the pnpm project’s preference is for what Node should do. Should Node’s installation write a `/usr/bin/local/pnpm` binary, and if so, what should that binary do (symlink to Corepack or run some other command).

**Comment by zkochan** [comment]
I personally prefer to install pnpm with the standalone script. This way pnpm can install node.js and manage node.js versions. The executable version of pnpm is also slightly faster.

I created a poll to see what the users prefer: https://twitter.com/pnpmjs/status/1764788568398897439

I personally don't want to be involved in the corepack discussions anymore. They are too heated.

**Reply by zkochan** [reply]
What choice do we have?

If we say we don't want to be included then 2 things may happen:

* Corepack gets cancelled. Yarn is also not included. Yarn team gets mad at us. npm gets the advantage of being the "standard" package manager
* Yarn is included. So now both Yarn and npm are shipped with node.js and pnpm is a "non-standard" package manager.

**Reply by MylesBorins** [reply]
I'd say there are other options

1. pnpm ships directly in node.js (vendor like npm)
2. pnpm works with Node.js team on streamlined bootstrap process that isn't based on corepack

I'd be willing to spend some time to help navigate this (or have someone from the npm team pair on it). If pnpm wants to find a path to shipping in Node.js without corepack they is possible, and it is possible without undermining the goals of the yarn team imho.

**Reply by GeoffreyBooth** [reply]
Myles’ option 2 is proposed here: https://github.com/nodejs/node/issues/51931.

We just ruled out option 1 in https://github.com/nodejs/node/pull/51918, though of course we can always change prior decisions.

**Reply by zkochan** [reply]
I have created another poll: https://twitter.com/pnpmjs/status/1764951109469098185

**Reply by GeoffreyBooth** [reply]
In the new poll, if the non-Corepack option was meant to suggest https://github.com/nodejs/node/issues/51931, that proposal wasn’t to have Node include a fixed version of pnpm. I was thinking we would download either the latest or the latest major version of pnpm.

Also for https://github.com/nodejs/node/issues/51931, I would expect pnpm to tell us what command they want to be used for installation. I used `npm install` in my experiment because that’s what I’m familiar with, but that was just to prove that the concept could work. I’m happy to change it to something else, provided that whatever pnpm suggests works in all the platforms that Node supports. We can also change it to use Corepack to install pnpm if pnpm wants to be managed by Corepack via this method.

**Reply by GeoffreyBooth** [reply]
Hi, it looks like the polls are complete.

<img width="579" alt="image" src="https://github.com/pnpm/pnpm/assets/456802/054017b2-75b4-4028-abbe-562c5910e7a6">

<img width="580" alt="image" src="https://github.com/pnpm/pnpm/assets/456802/a62f8b19-4db3-4686-b8e8-41d04de40a5c">

I’m not really sure what to make of the second one, since “a script to download the latest version” wasn’t an option; and based on the first poll that might very well have been preferred by people?

Or is what I should infer out of this that you don’t have a strong preference and you’re fine with the Node team deciding?
